### ISSUE TITLE
Compatibility with pg-promise?

### ISSUE THREAD

**Issue by vinc** [Joris-van-der-Wel/node-pg-large-object#4]
Hello,

I'm using `pg-promise` on a project exposing an API for streaming resized images from a database, and I'd like to use `pg-large-object` to stream a file from the database to the disk.

I got it working with the regular `pg` but I don't really see if it's possible to do it with `pg-promise` instead, more specifically how to create a promise from `openAndReadableStream`?

Here's some kind of pseudo code missing the creation of the promise:

    db.tx(function(t) {
      const man = new LargeObjectManager(db); // FIXME: `pg-promise` don't expose `client'
      const len = 16384;

      man.openAndReadableStream(oid, len, function(err, size, stream) {
        if (err) {
          done(err); // FIXME: `pg-promise` don't expose `done()`
          console.error('Unable to read the given lo', err);
          return next(err);
        }

        // Stream to temporary file
        const token = crypto.randomBytes(24).toString('hex');
        const cache = fs.createWriteStream(req.image.path + token);

        stream.pipe(cache);

        stream.on('end', function() {
          fs.link(req.image.path + token, req.image.path, function(err) {
            next();
          });
        });
      });
    }).then(function(stream) {
      // TODO: this is where writing to file should probably be happening
    }).catch(function(err) {
      // TODO: error handling here
    });

Thanks!

**Comment by Joris-van-der-Wel** [comment]
This library is built around the [pg](https://www.npmjs.com/package/pg) library, so it expects an `pg.Client` object.

Here are two solutions that you could use:

# 1. Write a small adapter
This library only needs the `query()` method of `pg.Client`. So we could write a small wrapper around pg-promise. This is easy although it is more fragile than the second solution

```javascript
'use strict';
const fs = require('fs');
const {LargeObjectManager} = require('pg-large-object');
const Promise = require('bluebird');
const pgp = require('pg-promise')({promiseLib: Promise});

Promise.promisifyAll(LargeObjectManager.prototype, {multiArgs: true});

const clientAdapter = tx => ({
    query: (obj, callback) => tx.query(obj).then(rows => ({rows})).asCallback(callback),
});

const db = pgp('postgres://joris:abc@localhost:5432/test');

db.tx(tx => {
    const man = new LargeObjectManager(clientAdapter(tx));

    return man.createAndWritableStreamAsync().then(([oid, stream]) => {
        console.log('Here is a new large object:', oid);

        const fileStream = fs.createReadStream('streep.jpg');
        fileStream.pipe(stream);

        return new Promise(resolve => stream.on('finish', resolve));
    });
}).then(() => console.log('All done!'));

```
If this works properly, I might add the adapter to this library

# 2. Obtain pg.Client
It looks like the only place where pg-promise exposes a `pg.Client` is in the `db.connect` method. So you would have to do something like this:

```javascript
'use strict';
const fs = require('fs');
const pgp = require('pg-promise')();
const {LargeObjectManager} = require('pg-large-object');
const Promise = require('bluebird');

Promise.promisifyAll(LargeObjectManager.prototype, {multiArgs: true});

const db = pgp('postgres://joris:abc@localhost:5432/test');

db.connect().then(connection => Promise.resolve()
    .then(() => {
        return connection.tx(t => {
            const man = new LargeObjectManager(connection.client);

            return man.createAndWritableStreamAsync().then(([oid, stream]) => {
                console.log('Here is a new large object:', oid);

                const fileStream = fs.createReadStream('streep.jpg');
                fileStream.pipe(stream);

                return new Promise(resolve => stream.on('finish', resolve));
            });
        });
    })
    .then(() => console.log('All done!'))
    .finally(() => connection.done())
);
```

**Comment by vinc** [comment]
Wow cool, I implemented your first suggestion to read from the database and it works perfectly!

https://github.com/vinc/infra.stru.ctu.re/commit/931a5aa40eb702d343457b63e6c4f9918f4deb69

(the code is a bit ugly, it's a quick prototype hacked together to extract that functionality from a larger rails app)

I spent some time earlier today looking at the internal of `pg-promise` to see how we could use `Client` but I didn't go as far as finding your second solution!

Cheers!

**Comment by vitaly-t** [comment]
The second solution looks cleaner to me. It is one of those special cases for which I designed the library to expose the `Client` object.

Under a normal scenario one doesn't use the `Client`, this is why it is not available, to avoid confusion :wink:

A little improved solution, and more correct one:

```js
var connection;

db.connect()
    .then(cn => {
        connection = cn;
        return cn.tx(t => {
            const man = new LargeObjectManager(cn.client);
            return man.createAndWritableStreamAsync()
                .then(([oid, stream]) => {
                    console.log('Here is a new large object:', oid);
                    const fileStream = fs.createReadStream('streep.jpg');
                    fileStream.pipe(stream);
                    return new Promise(resolve => stream.on('finish', resolve));
                });
        });
    })
    .then(data => {
        // SUCCESS
    })
    .catch(error => {
        // ERROR
    })
    .finally(() => {
        if (connection) {
            connection.done();
        }
    });
```

**Comment by vinc** [comment]
Thank you @vitaly-t, I implemented your solution: https://github.com/vinc/infra.stru.ctu.re/commit/127cf84e22440b0102aa5c27498b8726b517d158

So I removed this:

```javascript
const clientAdapter = t => ({
  query: (obj, callback) => t.query(obj).then(rows => ({rows})).asCallback(callback)
});
```

And replaced this:

```javascript
const streamFile = function(req, res, next) {
  db.tx(function(t) {
    return t.one(oidSQL(req.params), req.params).then(function(image) {
      const man = new LargeObjectManager(clientAdapter(t));
      return man.openAndReadableStreamAsync(image.oid).then(function([size, stream]) {
        const temp = fs.createWriteStream(req.tempCache);

        stream.pipe(temp);

        return new Promise(function(resolve) {
          stream.on('end', function() {
            fs.rename(req.tempCache, req.originalCache, function(err) {
              if (err) {
                return next(err);
              }
              resolve();
            });
          });
        });
      });
    });
  }).then(next).catch(next);
};
```

With this:

```javascript
const streamFile = function(req, res, next) {
  var connection;

  db.connect().then(cn => {
    connection = cn;
    return cn.tx(t => {
      return t.one(oidSQL(req.params), req.params).then(function(image) {
        const man = new LargeObjectManager(cn.client);
        return man.openAndReadableStreamAsync(image.oid).then(function([size, stream]) {
          const temp = fs.createWriteStream(req.tempCache);

          stream.pipe(temp);

          return new Promise(function(resolve) {
            stream.on('end', function() {
              fs.rename(req.tempCache, req.originalCache, function(err) {
                if (err) {
                  return next(err);
                }
                resolve();
              });
            });
          });
        });
      });
    }).then(next).catch(next).finally(function() {
      if (connection) {
        connection.done();
      }
    });
  });
};
```

Cheers!

**Comment by vitaly-t** [comment]
@Joris-van-der-Wel if you are interested in either migrating or better support of [pg-promise], or even simply providing an example that uses it, you already have a good starting example :wink:

And if you want a better one - I will help you with that :wink:

What makes it worthwhile is the simplest and most reliable support for transactions that [pg-promise] provides, makes the solution so much shorter and nicer :wink:

[pg-promise]:https://github.com/vitaly-t/pg-promise

**Comment by Joris-van-der-Wel** [comment]
Hey, I've added promise support in a branch, check it out: https://github.com/Joris-van-der-Wel/node-pg-large-object/tree/promise

I did not want to change around things too drastically because this library is 3 years old. So I made a few tweaks:

* Add a `fooBarAsync` variant for all methods that take a callback. So you can either use `lo.size(callback)` or `promise = lo.sizeAsync()`
* Support for passing a pg-promise `Database` or `Task` object to the LargeObjectManager constructor (a database transaction is a `Task`)
* Add some examples that demonstrate using this library with pg-promise

The old API will continue to work as it did before.

**Comment by vitaly-t** [comment]
@Joris-van-der-Wel  for the new examples, I would advise adding `.then` and `.catch` in the end of the transaction, as normal practice when using promises.

**Comment by Joris-van-der-Wel** [comment]
Good point, i've added them

**Comment by Joris-van-der-Wel** [comment]
Released in v2.0.0

**Comment by vitaly-t** [comment]
@Joris-van-der-Wel 

Your examples of using `pg` with callbacks are not very valid, as they do not execute `ROLLBACK` in case of failure. This is what `pg-promise` does automatically, by the way :wink:
